libs/s25main/Replay.cpp

```diff
@@ -32,28 +33,76 @@ uint16_t Replay::GetVersion() const
 {
     /// Version des Replay-Formates
     /// Search for "TODO(Replay)" when increasing this (breaking Replay compatibility)
-    return 6;
+    return 7;
 }
 
 //////////////////////////////////////////////////////////////////////////
 
-Replay::Replay() : random_init(0), isRecording(false), lastGF_(0), last_gf_file_pos(0), mapType_(MapType::OldMap) {}
+Replay::Replay() : random_init(0), isRecording_(false), lastGF_(0), lastGfFilePos_(0), mapType_(MapType::OldMap) {}
 
-Replay::~Replay()
-{
-    StopRecording();
-}
+Replay::~Replay() = default;
 
 void Replay::Close()
 {
-    file.Close();
+    file_.Close();
+    uncompressedDataFile_.reset();
+    isRecording_ = false;
+    filepath_.clear();
     ClearPlayers();
 }
 
-void Replay::StopRecording()
+bool Replay::StopRecording()
 {
-    file.Close();
-    isRecording = false;
+    if(!isRecording_)
+        return true;
+    const unsigned replayDataSize = file_.Tell();
+    isRecording_ = false;
+    file_.Close();
+
+    BinaryFile file;
+    if(!file.Open(filepath_, OpenFileMode::OFM_READ))
+        return false;
+    try
+    {
+        TmpFile tmpReplayFile(".rpl");
+        file.Seek(0, SEEK_SET);
+        tmpReplayFile.close();
+        BinaryFile compressedReplay;
+        compressedReplay.Open(tmpReplayFile.filePath, OpenFileMode::OFM_WRITE);
+
+        // Copy header data uncompressed
+        std::vector<char> data(lastGfFilePos_);
+        file.ReadRawData(data.data(), data.size());
+        compressedReplay.WriteRawData(data.data(), data.size());
+        const auto lastGF = file.ReadUnsignedInt();
+        RTTR_Assert(lastGF == lastGF_);
+        compressedReplay.WriteUnsignedInt(lastGF);
+        file.ReadUnsignedChar(); // Ignore compressed flag
+        compressedReplay.WriteUnsignedChar(1);
+
+        // Read and compress remaining data
+        const auto uncompressedSize = replayDataSize - file.Tell();
+        data.resize(uncompressedSize);
+        file.ReadRawData(data.data(), data.size());
+        data = CompressedData::compress(data);
+        // If the compressed data turns out to be larger than the uncompressed one, bail out
+        // This can happen for very short replays
+        if(data.size() >= uncompressedSize)
+            return true;
+        compressedReplay.WriteUnsignedInt(uncompressedSize);
+        compressedReplay.WriteUnsignedInt(data.size());
+        compressedReplay.WriteRawData(data.data(), data.size());
+
+        // All done. Replace uncompressed replay
+        compressedReplay.Close();
+        file.Close();
+        boost::filesystem::rename(tmpReplayFile.filePath, filepath_);
```

Comment:
does this overwrite existing files?

Reply:
Yes and it should as it is meant to overwrite the just created replay: https://www.boost.org/doc/libs/1_45_0/libs/filesystem/v3/doc/reference.html#rename